web/impact/impact/v1/classes/option_analysis.py

```diff
@@ -96,38 +140,164 @@ def calc_needs(self, option_name):
         expected_count = self.option_spec.count
         return {expected_count - k: v for (k, v) in counts.items()}
 
-    def completed_feedbacks(self):
-        return feedbacks_for_judging_round(
-            self.judging_round, self.apps).filter(
-                feedback_status=JUDGING_FEEDBACK_STATUS_COMPLETE)
-
     def calc_capacity(self, option_name):
         commitments = JudgeRoundCommitment.objects.filter(
             judging_round=self.judging_round)
-        total_capacity = self.helper.total_capacity(
-            commitments=commitments,
-            option_name=option_name)
-        remaining_capacity = self.helper.remaining_capacity(
-            commitments,
+        criteria_function = self.criterion_total_functions[
+            (self.option_spec.criterion.type, self.option_spec.criterion.name)
+        ]['function']
+        total_capacity = (
+            criteria_function(option_name)
+            if criteria_function is not None
+            else self.helper.total_capacity(
+                commitments=commitments,
+                option_name=option_name)
+        )
+        remaining_capacity = self.remaining_capacity(
             self.application_counts,
+            self.option_spec,
             option_name)
         return {
             "total_capacity": total_capacity,
             "remaining_capacity": remaining_capacity,
         }
 
+    def general_criterion_total_capacity(self, option):
+        option_name = self.option_spec.criterion.name
+        field = self.criterion_total_functions[
+            (self.option_spec.criterion.type, option_name)
+        ]['filter_field']
+
+        if option_name == "gender":
+            option = self.gender_match_dict.get(option)
+
+        if self.criterion_total_capacities.get(option_name) is None:
+            capacities = JudgeRoundCommitment.objects.filter(
+                judging_round=self.judging_round) \
+                .values(field) \
+                .annotate(total=Sum("capacity"))
+            result = {
+                cap[field]: cap["total"]
+                for cap in capacities}
+            self.criterion_total_capacities[option_name] = result
+
+        key_exists = self.criterion_total_capacities[option_name].get(option)
+        return 0 if not key_exists else self.criterion_total_capacities[
+            option_name][option]
+
+    def remaining_capacity(self, assignment_counts, option_spec, option):
+        if self.judge_to_capacity_cache is None:
+            self.judge_to_capacity_cache = JudgeRoundCommitment.objects.filter(
+                judging_round=self.judging_round).values(
+                    "judge_id",
+                    self.industry_judge_field,
+                    self.program_judge_field,
+                    self.gender_judge_field,
+                    self.role_judge_field,
+                    "capacity"
+                ).annotate(
+                    industry=F(self.industry_judge_field),
```

Comment:
These can come from criterion helpers. In this case (and in the similar case below) we'd do this by getting a dict of `{field_name: field}` from each criterion helper and updating a kwargs dict, which we pass to the `annotate` function. Something like
```
fields = ["judge_id"] + [helper.field() for helper in self.criterion_helpers]
annotations = {criterion_helper.annotation() for criterion in self.criterion_helpers}
self.judge_to_capacity_cache = JudgeRoundCommitment.objects.filter(
    judging_round=self.judging_round).values(*fields).
       annotate(**annotations)
```